storage: TestRaftTransportIndependentRanges failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/be35db9bddf0739a3971d4095ce19d896e251738

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestRaftTransportIndependentRanges PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1262716&tab=buildLog

```

W190426 06:31:42.606634 633493 storage/raft_transport.go:436  no handler found for store 2 in response range_id:13 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport_test.go:71: channelServer broken range" transaction_restart:NONE origin_node:0 detail:<> now:<> > > 
W190426 06:31:42.611780 633493 storage/raft_transport.go:436  no handler found for store 2 in response range_id:13 from_replica:<node_id:1 store_id:1 replica_id:1 > to_replica:<node_id:2 store_id:2 replica_id:2 > union:<error:<message:"storage/raft_transport_test.go:71: channelServer broken range" transaction_restart:NONE origin_node:0 detail:<> now:<> > > 
panic: locally configured maximum clock offset (1ns) does not match that of node 127.0.0.1:38873 (500ms)

goroutine 633547 [running]:
github.com/cockroachdb/cockroach/pkg/rpc.(*HeartbeatService).Ping(0xc001d882d0, 0x3556420, 0xc000d55770, 0xc002090a10, 0xc001d882d0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.go:134 +0x71b
github.com/cockroachdb/cockroach/pkg/rpc._Heartbeat_Ping_Handler.func1(0x3556420, 0xc000d55770, 0x2ccdcc0, 0xc002090a10, 0x2ccdcc0, 0xc002090a10, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.pb.go:227 +0x86
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x3556420, 0xc000d55770, 0x2ccdcc0, 0xc002090a10, 0xc0022fb1e0, 0xc0022fb200, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xa08
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func4(0x3556420, 0xc000d55770, 0x2ccdcc0, 0xc002090a10, 0xc0022fb1e0, 0xc0022fb200, 0x2a969a0, 0x507d9d8, 0x2d84de0, 0xc000eaef00)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:252 +0xac
github.com/cockroachdb/cockroach/pkg/rpc._Heartbeat_Ping_Handler(0x2a61fe0, 0xc001d882d0, 0x3556420, 0xc000d55770, 0xc0020909a0, 0xc00dd7e0b0, 0x0, 0x0, 0xc000eaef00, 0xc0016603b0)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/heartbeat.pb.go:229 +0x158
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc00f5ec000, 0x357ac00, 0xc000290000, 0xc000eaef00, 0xc001d88360, 0x4dafda0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1011 +0x4cd
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc00f5ec000, 0x357ac00, 0xc000290000, 0xc000eaef00, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1249 +0x1311
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00028aa10, 0xc00f5ec000, 0x357ac00, 0xc000290000, 0xc000eaef00)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:680 +0x9f
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xa1

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00190ff00, 0x2e235a5, 0x22, 0x2f387f8, 0x922a01)
	/usr/local/go/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc00049e200)
	/usr/local/go/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc00049e200, 0xc000661ba0)
	/usr/local/go/src/testing/testing.go:827 +0xbf
testing.runTests(0xc0002ae3a0, 0x4de29c0, 0x230, 0x230, 0x1771c14)
	/usr/local/go/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc00024a200, 0x0)
	/usr/local/go/src/testing/testing.go:1034 +0x165
github.com/cockroachdb/cockroach/pkg/storage_test.TestMain(0xc00024a200)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/main_test.go:57 +0x1e1
main.main()
	_testmain.go:1170 +0x13d

goroutine 6 [syscall, 2 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 7 [chan receive]:
```

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -834,73 +834,79 @@ func (ctx *Context) runHeartbeat(
 		select {
 		case <-redialChan:
 			return grpcutil.ErrCannotReuseClientConn
-		case <-ctx.Stopper.ShouldStop():
+		case <-ctx.Stopper.ShouldQuiesce():
 			return nil
 		case <-heartbeatTimer.C:
 			heartbeatTimer.Read = true
 		}
 
-		// We re-mint the PingRequest to pick up any asynchronous update to clusterID.
-		clusterID := ctx.ClusterID.Get()
-		request := &PingRequest{
-			Addr:           ctx.Addr,
-			MaxOffsetNanos: maxOffsetNanos,
-			ClusterID:      &clusterID,
-			NodeID:         conn.remoteNodeID,
-			ServerVersion:  ctx.version.ServerVersion,
-		}
+		if err := ctx.Stopper.RunTaskWithErr(ctx.masterCtx, "rpc heartbeat", func(goCtx context.Context) error {
+			// We re-mint the PingRequest to pick up any asynchronous update to clusterID.
+			clusterID := ctx.ClusterID.Get()
+			request := &PingRequest{
+				Addr:           ctx.Addr,
+				MaxOffsetNanos: maxOffsetNanos,
+				ClusterID:      &clusterID,
+				NodeID:         conn.remoteNodeID,
+				ServerVersion:  ctx.version.ServerVersion,
+			}
 
-		var response *PingResponse
-		sendTime := ctx.LocalClock.PhysicalTime()
-		err := contextutil.RunWithTimeout(ctx.masterCtx, "rpc heartbeat", ctx.heartbeatTimeout,
-			func(goCtx context.Context) error {
-				// NB: We want the request to fail-fast (the default), otherwise we won't
-				// be notified of transport failures.
-				var err error
-				response, err = heartbeatClient.Ping(goCtx, request)
-				return err
+			var response *PingResponse
+			sendTime := ctx.LocalClock.PhysicalTime()
+			err := contextutil.RunWithTimeout(goCtx, "rpc heartbeat", ctx.heartbeatTimeout,
+				func(goCtx context.Context) error {
+					// NB: We want the request to fail-fast (the default), otherwise we won't
+					// be notified of transport failures.
+					var err error
+					response, err = heartbeatClient.Ping(goCtx, request)
+					return err
+				})
+
+			if err == nil {
+				err = errors.Wrap(
+					checkVersion(ctx.version, response.ServerVersion),
+					"version compatibility check failed on ping response")
+			}
+
+			if err == nil {
+				everSucceeded = true
+				receiveTime := ctx.LocalClock.PhysicalTime()
+
+				// Only update the clock offset measurement if we actually got a
+				// successful response from the server.
+				pingDuration := receiveTime.Sub(sendTime)
+				maxOffset := ctx.LocalClock.MaxOffset()
+				if maxOffset != timeutil.ClocklessMaxOffset &&
+					pingDuration > maximumPingDurationMult*maxOffset {
+
+					request.Offset.Reset()
+				} else {
+					// Offset and error are measured using the remote clock reading
+					// technique described in
+					// http://se.inf.tu-dresden.de/pubs/papers/SRDS1994.pdf, page 6.
+					// However, we assume that drift and min message delay are 0, for
+					// now.
+					request.Offset.MeasuredAt = receiveTime.UnixNano()
+					request.Offset.Uncertainty = (pingDuration / 2).Nanoseconds()
+					remoteTimeNow := timeutil.Unix(0, response.ServerTime).Add(pingDuration / 2)
+					request.Offset.Offset = remoteTimeNow.Sub(receiveTime).Nanoseconds()
+				}
+				ctx.RemoteClocks.UpdateOffset(ctx.masterCtx, target, request.Offset, pingDuration)
+
+				if cb := ctx.HeartbeatCB; cb != nil {
+					cb()
+				}
+			}
+			conn.heartbeatResult.Store(heartbeatResult{
+				everSucceeded: everSucceeded,
+				err:           err,
 			})
+			conn.setInitialHeartbeatDone()
 
-		if err == nil {
-			err = errors.Wrap(
-				checkVersion(ctx.version, response.ServerVersion),
-				"version compatibility check failed on ping response")
+			return nil
+		}); err != nil {
+			return err
 		}
-
-		if err == nil {
-			everSucceeded = true
-			receiveTime := ctx.LocalClock.PhysicalTime()
-
-			// Only update the clock offset measurement if we actually got a
-			// successful response from the server.
-			pingDuration := receiveTime.Sub(sendTime)
-			maxOffset := ctx.LocalClock.MaxOffset()
-			if maxOffset != timeutil.ClocklessMaxOffset &&
-				pingDuration > maximumPingDurationMult*maxOffset {
-
-				request.Offset.Reset()
-			} else {
-				// Offset and error are measured using the remote clock reading
-				// technique described in
-				// http://se.inf.tu-dresden.de/pubs/papers/SRDS1994.pdf, page 6.
-				// However, we assume that drift and min message delay are 0, for
-				// now.
-				request.Offset.MeasuredAt = receiveTime.UnixNano()
-				request.Offset.Uncertainty = (pingDuration / 2).Nanoseconds()
-				remoteTimeNow := timeutil.Unix(0, response.ServerTime).Add(pingDuration / 2)
-				request.Offset.Offset = remoteTimeNow.Sub(receiveTime).Nanoseconds()
-			}
-			ctx.RemoteClocks.UpdateOffset(ctx.masterCtx, target, request.Offset, pingDuration)
-
-			if cb := ctx.HeartbeatCB; cb != nil {
-				cb()
-			}
-		}
-		conn.heartbeatResult.Store(heartbeatResult{
-			everSucceeded: everSucceeded,
-			err:           err,
-		})
-		conn.setInitialHeartbeatDone()
 
 		heartbeatTimer.Reset(ctx.heartbeatInterval)
 	}

